src/codegen.cpp

```diff
@@ -2033,14 +2063,22 @@ static void jl_add_method_root(jl_codectx_t &ctx, jl_value_t *val)
     else {
         size_t rlen = jl_array_dim0(ctx.roots);
         for (size_t i = 0; i < rlen; i++) {
-            if (jl_array_ptr_ref(ctx.roots,i) == val) {
+            jl_value_t *rt = jl_array_ptr_ref(ctx.roots,i);
+            if (rt == val || jl_egal(rt, val)) {
                 JL_GC_POP();
-                return;
+                return rt;
             }
         }
+        jl_method_t *m = ctx.linfo->def.method;
+        int needlock = m->roots == ctx.roots;
+        if (needlock) JL_LOCK(&m->writelock);
```

Comment:
I don't think we don't really want to be mutating `m` in this low-level function.